executors: Add high-level documentation on firecracker (#51306)
sourcegraph/sourcegraph: doc/admin/executors/firecracker.md

# Firecracker

> [Executors](./index.md), by design, are services that run arbitrary code supplied by a user. The executor jobs produced by precise code intelligence [auto-indexing](../../code_navigation/explanations/auto_indexing.md) and [server-side batch changes](../../batch_changes/explanations/server_side.md) are built to invoke _templated_ execution plans, where some parts of execution may invoke code configured by a Sourcegraph administrator or user. Generating a precise index requires invoking an indexer for that language. Batch changes are configured to run arbitrary tooling over a the contents of a repository.

Reviewer: ```suggestion
[Executors](./index.md), by design, are services that run arbitrary code supplied by a user. The executor jobs produced by precise code intelligence [auto-indexing](../../code_navigation/explanations/auto_indexing.md) and [server-side batch changes](../../batch_changes/explanations/server_side.md) are built to invoke _templated_ execution plans, where some parts of execution may invoke code configured by a Sourcegraph administrator or user. Generating a precise index requires invoking an indexer for that language. Batch changes are configured to run arbitrary tooling over the contents of a repository.
```

Jobs handled by the executor are defined as a series of Docker (or Docker-like) image invocations sharing the same filesystem workspace initialized with the contents of the repository being indexed or modified. When the executor process is running on a raw compute node (specifically, not using the Kubernetes runner in-cluster), the executor can create a [Firecracker "MicroVM"](https://firecracker-microvm.github.io/) for every job and run the containers in a _completely isolated_ manner. The Docker (or Docker-like) daemon running in a MicroVM is isolated from the host, hence each job is also isolated from one another.

See [this architecture diagram](./index.md#firecracker) detailing Firecracker isolation.

## When to use

Using Firecracker is our **most secure** isolation method, but it is not a **necessary** isolation method. Most users will be fine running containers on the executor host, or deploying bundled executors via Kubernetes jobs. Firecracker isolation was created when our design constrains included multi-tenant environments, and is likely overkill for any on-premise Sourcegraph instance administrated by a single company.

> For companies with very high security consciousness, Firecracker isolation is still an option. See the following section for installation instructions as well as deployment caveats. Note that Firecracker relies on some specific kernel extensions to run, which are only available on some classes of compute on Cloud providers such as AWS and GCP.

Reviewer 2: I was slightly confused by the phrase "See the following section for installation instructions as well as deployment caveats." I was expecting a list of links to these (took me time to realized this is referencing the next sections).

Potentially some suggestions to make this clearer.

#### Remove the phrase.

```suggestion
For companies with very high security consciousness, Firecracker isolation is still an option. 

> Note: Firecracker relies on some specific kernel extensions to run, which are only available on some classes of compute on Cloud providers such as AWS and GCP.
```

#### Add link

```suggestion
For companies with very high security consciousness, Firecracker isolation is still an option. See the [How to use](firecracker.md#how-to-use) for installation instructions and deployment caveats.

> Note: Firecracker relies on some specific kernel extensions to run, which are only available on some classes of compute on Cloud providers such as AWS and GCP.
```

## How to use

To use Firecracker, the host machine has to support KVM. When deploying on an AWS, that means the compute node must run on a [metal instance (`.metal`)](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-types.html). When deploying on GCP, that means the compute node must have [nested virtualization enabled](https://cloud.google.com/compute/docs/instances/nested-virtualization/enabling).

See [deploying Executors binary](./deploy_executors_binary.md) for additional information on configuring Linux Machines.

## Known caveats

> We configure iptables to prevent Firecracker-isolated processes from talking on local networks (providing network-level isolation). They can talk to DNS and Sourcegraph only, which prevents users from talking to a 10 range IP.

Reviewer: ```suggestion
We configure iptables to prevent Firecracker-isolated processes from talking on [Private IPv4 Addresses](https://en.wikipedia.org/wiki/Private_network#Private_IPv4_addresses) networks (providing network-level isolation). They can talk to DNS and Sourcegraph only, which prevents users from talking to a 10.x.x.x, 172.x.x.x, or 192.168.x.x range IP.
```

For companies with very high security consciousness, Firecracker isolation is still an option. See the [How to use](firecracker.md#how-to-use) for installation instructions and deployment caveats.

> Note: Firecracker relies on some specific kernel extensions to run, which are only available on some classes of compute on Cloud providers such as AWS and GCP.

## How to use

To use Firecracker, the host machine has to support KVM. When deploying on an AWS, that means the compute node must run on a [metal instance (`.metal`)](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-types.html). When deploying on GCP, that means the compute node must have [nested virtualization enabled](https://cloud.google.com/compute/docs/instances/nested-virtualization/enabling).

See [deploying Executors binary](./deploy_executors_binary.md) for additional information on configuring Linux Machines.

## Known caveats

> We configure iptables to prevent Firecracker-isolated processes from talking on [Private IPv4 Addresses](https://en.wikipedia.org/wiki/Private_network#Private_IPv4_addresses) (providing network-level isolation). They can talk to DNS and Sourcegraph only, which prevents users from talking to a 10.x.x.x, 172.x.x.x, or 192.168.x.x range IP.

Reviewer: ```suggestion
We [configure iptables](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/cmd/executor/internal/run/install.go?L229-255) to prevent Firecracker-isolated processes from talking on [Private IPv4 Addresses](https://en.wikipedia.org/wiki/Private_network#Private_IPv4_addresses) (providing network-level isolation). They can talk to DNS and Sourcegraph only, which prevents users from talking to a 10.x.x.x, 172.x.x.x, or 192.168.x.x range IP.
```

Adds in a Sourcegraph link to the code where this is configured since these docs are more-than-likely used during executor development rather than end-users and this way the next person who reads this and goes "huh?" can find out how it works.
